lib/guardian_extensions.rb

```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module GuardianExtensions
+  def can_see_topic?(topic, hide_deleted = true)
+    if SiteSetting.encrypt_enabled? && topic.private_message? && topic.is_encrypted?
+      authenticated? && topic.all_allowed_users.where(id: @user.id).exists?
+    else
+      true
+    end && super
+  end
+
+  def can_see_post?(post)
+    can_see_topic?(post.topic) && super
```

Comment:
```suggestion
    can_see_topic?(post&.topic) && super
```

I believe `post` is allowed to be `nil` in this case too.

From core:

```ruby
  def can_see_post?(post)
    return false if post.blank?
    # ...
  end
```